### ISSUE TITLE
[Feature]: Add a metric-registration user interface for training/validation

### ISSUE THREAD

**Issue by debrevitatevitae** [GroupiSP/probreg#5]
### Summary

There is currently no way for users to register custom metrics to be computed during training/validation and forwarded to `EventSink`/`ExperimentTracker` implementations.

Today, `probreg.jax.supervised.run_supervised` and `probreg.jax.evaluation.evaluate_loader` hardcode a single `"loss"` metric per epoch/validation pass. Pure NumPy metric functions already exist in `src/probreg/core/metrics.py` (`rmse`, `coverage`, `wsu`, `crps`, `point_crps`, `cdf`), but there is no registration mechanism to plug them (or JAX-native, batch-level metrics) into the training loop so their values flow into the `metrics: Mapping[str, float]` field of `TrainingEvent` and onward to sinks/trackers.

This limits observability: users must fork/monkeypatch the training runner to track anything beyond loss (e.g. RMSE, interval coverage, CRPS) during training.

### Proposed approach

Consulted with `.github/agents/python-scientific-architect.agent.md` for the design. Summary below.

**1. Rationale**

Two metric classes have genuinely different execution contexts, so they need two registration mechanisms behind one façade:

* **Batch metrics** are JAX-native, computed inside the traced `nnx.jit` step alongside the loss (cheap, no host sync per quantity), then reduced across batches on the host.
* **Epoch metrics** (`rmse`, `coverage`, `wsu`, `crps`) are pure NumPy and cannot be traced. They need *materialized* predictions/targets accumulated over a loader, then a single host-side call.

Unifying them under one callable `Protocol` would force either tracing NumPy or host-syncing every batch. Instead: two `Protocol`s, one `MetricSuite` container passed by composition — no changes to `TrainingEvent`, `EventSink`, or `ExperimentTracker`.

**2. Signatures and placement**

`core/metric_registry.py` (new, JAX-free — epoch side is backend-neutral):

```python
@dataclass(frozen=True, slots=True)
class MetricInputs:
    targets: NDArray[np.float64]
    mean: NDArray[np.float64]
    variance: NDArray[np.float64] | None = None
    lower: NDArray[np.float64] | None = None
    upper: NDArray[np.float64] | None = None
    metadata: Mapping[str, Any] = field(default_factory=dict)

class EpochMetric(Protocol):
    @property
    def name(self) -> str: ...
    def __call__(self, data: MetricInputs, /) -> float: ...
```

Thin frozen adapters (`RootMeanSquaredError()`, `IntervalCoverage(level=0.95)`, `WeightedSpread(coordinate_key="x")`) wrap the existing `core/metrics.py` functions — those stay untouched pure functions.

`jax/metrics.py` (new):

```python
class BatchMetric(Protocol):  # same shape as SupervisedLoss -> reusable
    def __call__(self, model, inputs, targets, sample_weight, key, training, /) -> jax.Array: ...

@dataclass(frozen=True, slots=True)
class BatchMetricSpec:
    name: str
    metric: BatchMetric
    reduce: Callable[[Sequence[float]], float] = _metric_mean

@dataclass(frozen=True, slots=True)
class MetricSuite:
    batch: tuple[BatchMetricSpec, ...] = ()
    epoch: tuple[EpochMetric, ...] = ()
    predictor: Predictor | None = None   # optional override; see below
    def __post_init__(self): ...  # reject duplicate/"loss" names
```

**Refinement (per review discussion):** rather than requiring callers to separately construct and pair a `Predictor` with their model/head, heads should own the mapping from raw outputs to `MetricInputs` directly, since that mapping is intrinsic to the head's output semantics (a Gaussian head always decodes to `mean`+`variance`; a quantile head always decodes to `lower`+`upper`). This avoids a class of "wrong predictor for this head" bugs and cuts API surface.

```python
# core/metric_registry.py — stays JAX-free
@runtime_checkable
class SupportsMetricInputs(Protocol):
    def produce_metric_inputs(self, batch: Batch, /) -> MetricInputs: ...
```

Any head (e.g. a future `GaussianHead` in `jax/`) implements this directly, converting its own JAX outputs to NumPy internally. `MetricSuite.predictor` becomes an **optional override** rather than a required field. Resolution order at call time: if `suite.predictor` is set, use it; else fall back to `model.produce_metric_inputs` if the model satisfies `SupportsMetricInputs`; else raise a clear `ValueError` if `suite.epoch` is non-empty but neither is available. `Predictor` is only invoked when `suite.epoch` is non-empty (zero memory cost otherwise), and the override is retained intentionally to still allow research use cases — e.g. evaluating epoch metrics on a transformed/ensembled prediction — without forcing every experiment into the head's default decoding.

**Wiring** (extend existing modules, no new runner):
- `make_evaluation_step(loss, *, metrics=())` folds specs into the JIT'd step.
- `evaluate_loader(..., metrics=MetricSuite())` reduces per-name and appends epoch metrics.
- `HeldOutValidation` gains a `metrics: MetricSuite = MetricSuite()` field (existing `metric_prefix` applies uniformly).
- `run_supervised(..., metrics: MetricSuite = MetricSuite())` mirrors this on the train loop.

**3. Merge path**

`run_supervised` builds `metrics = {"loss": ..., **batch_metrics, **epoch_metrics}` — `"loss"` is written first and reserved, so registrations can never shadow it. That single dict flows unchanged into `_record_metrics` and `_emit`, so sinks/trackers get new keys for free. Validation results are prefixed by `HeldOutValidation` as today, keeping `early_stopper` able to monitor any registered metric.

**4. Backward compatibility**

Defaults are empty tuples; with no registration the emitted mapping is byte-identical to `{"loss": ...}`. One deliberate break to flag: the JIT'd step would return `dict[str, jax.Array]` (always containing `"loss"`) instead of a bare scalar, affecting direct callers of `make_evaluation_step`/`make_train_step` — this should be called out in release notes, or gated so a scalar is still returned when `metrics == ()`.

**5. Testing considerations**

Property-based (Hypothesis), per repo convention of favoring mathematical properties over example-based tests:
- Empty suite ⇒ output equals current behaviour (no regression).
- Batch-mean reduction equals whole-dataset computation for equal-sized batches.
- Epoch metrics equal direct `core/metrics` calls on concatenated arrays.
- Duplicate names, or a name colliding with `"loss"`, raise at `MetricSuite` construction.
- All values are finite Python `float`s (NaN gate).
- Results are deterministic under a fixed PRNG key.
- Every registered name appears in `state.metric_history` and in exactly one `TrainingEvent` per epoch, correctly prefixed for validation.

### Alternatives considered

- A single unified `Metric` Protocol for both batch- and epoch-level metrics — rejected because it would force either tracing NumPy code inside `nnx.jit` (impossible) or host-syncing every batch for JAX-native metrics (defeats the performance purpose of batch-level metrics).
- Extending `SupervisedLoss` itself to return an auxiliary metrics dict — rejected as it conflates the loss-optimization contract with observability concerns and would break the existing `SupervisedLoss` Protocol signature for all current implementers.

### Contribution

- [ ] I would like to work on it
